applications/GeoMechanicsApplication/custom_elements/permeability_calculator.h

```diff
@@ -0,0 +1,89 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Richard Faasse
+//
+
+#pragma once
+
+#include "calculator.h"
+
+#include <custom_retention/retention_law.h>
+
+#include <utility>
+
+namespace Kratos
+{
+
+class PermeabilityCalculator : public Calculator
+{
+public:
+    class InputProvider
+    {
+    public:
+        InputProvider(std::function<const Properties&()>                         rElementProperties,
+                      std::function<const std::vector<RetentionLaw::Pointer>&()> rRetentionLaws,
+                      std::function<Vector()>                        rIntegrationCoefficients,
+                      std::function<Vector(const Variable<double>&)> rNodalValuesOfDtWaterPressure,
+                      std::function<Geometry<Node>::ShapeFunctionsGradientsType()> rShapeFunctionGradients)
+            : mGetElementProperties(std::move(rElementProperties)),
+              mGetRetentionLaws(std::move(rRetentionLaws)),
+              mGetIntegrationCoefficients(std::move(rIntegrationCoefficients)),
+              mGetNodalValues(std::move(rNodalValuesOfDtWaterPressure)),
+              mGetShapeFunctionGradients(std::move(rShapeFunctionGradients))
+        {
+        }
+
+        [[nodiscard]] const Properties& GetElementProperties() const
+        {
+            return mGetElementProperties();
+        }
```

Comment:
Since the members of nested class `PermeabilityCalculator::InputProvider` are not template members, I would suggest to move their implementations to `permeability_calculator.cpp`. For instance, `GetElementProperties` would require two changes. First remove the implementation from the header file:
```suggestion
        [[nodiscard]] const Properties& GetElementProperties() const;
```

Then move its definition to `permeability_calculator.cpp`:
```cpp
const Properties& PermeabilityCalculator::InputProvider::GetElementProperties() const
{
    return mGetElementProperties();
}
```

Needless to say that the other members require similar changes.

Reply:
Done